### ISSUE TITLE
py_sig_re dropped without a deprecation warning

### ISSUE THREAD

**Issue by j-carson** [sphinx-doc/sphinx#12303]
### Describe the bug

autodoc_pydantic is using py_sig_re to parse pydantic models (which are just python objects) and it disappeared in 7.3 without a deprecation warning

### How to Reproduce

This is where the function is used

https://github.com/mansenfranzen/autodoc_pydantic/blob/afdec966659d994c585d8a7f0df27b7eb4212813/sphinxcontrib/autodoc_pydantic/directives/directives.py#L104

### Environment Information

```text
Sphinx 7.3.6
autodoc_pydantic
```

### Additional context
```
Running Sphinx v7.3.6

Extension error:
Could not import extension sphinxcontrib.autodoc_pydantic (exception: cannot import name 'py_sig_re' from 'sphinx.domains.python'
```

**Comment by AA-Turner** [comment]
cc: @mansenfranzen

We can make a patch release to reinstate ``py_sig_re``. I would note, though, that it is a private object -- it is not documented anywhere and the regular expression contained therein is subject to change without notice (and indeed, has done so).

It would be better to find a supported and documented method for ``autodoc_pydantic`` to achieve the same outcome.

A

**Comment by mansenfranzen** [comment]
@j-carson thanks for taking care with this issue! 

@AA-Turner it would be great if you could re-import the regular expression to allow older versions of autodoc_pydantic to work with the newest sphinx versions. 

For the future, I would also like to rely on a public function/API instead of using the private regex. Are you aware of any? If not, I would be willing to provide a PR adding the rather narrow functionality wrapped in a publicly facing function.

**Comment by AA-Turner** [comment]
62c3bad will be included in Sphinx 7.3.7.

For the second, I will have a think. Perhaps it could be as simple as splitting the first few lines from ``handle_signature`` and returning some custom type:

```python
def split_object_signature(sig: str) -> _SigType:
    m = py_sig_re.match(sig)
    if m is None:
        raise ValueError
    prefix, name, tp_list, arglist, retann = m.groups()
    return _SigType(...)
```

**Comment by AA-Turner** [comment]
[Sphinx 7.3.7](https://pypi.org/project/Sphinx/7.3.7/) has been released with fixes.

A
